### PR TITLE
Package emscripten

### PR DESCRIPTION
**Pull request by Ludea**

### CODE REVIEW

**Comment by Ludea** [comment]
Don't merge ! I get a warning on LLVM version.
Incoming PR comes to update it.

**Comment by Ludea** [comment]
Weird, it needs llvm 12.0, which is not release out

File: packages/emscripten/build.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
+TERMUX_PKG_HOMEPAGE=https://emscripten.org/
+TERMUX_PKG_DESCRIPTION="Emscripten is a complete compiler toolchain to WebAssembly"
+TERMUX_PKG_LICENSE="MIT"
+TERMUX_PKG_VERSION=2.0.9
+TERMUX_PKG_SRCURL=https://github.com/emscripten-core/emscripten/archive/${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=9e54b33da5025270221b38fdc603191b8f22afa038b72c6dcfaa6423888746f5
+TERMUX_PKG_DEPENDS="python, llvm, nodejs, binaryen"
+
+termux_step_post_get_source() { 
+	export EM_BINARYEN_ROOT=$TERMUX_PREFIX/usr/bin
```
Comment (Ludea) [inline packages/emscripten/build.sh:10]: So I need to set the full path of binaryen's binaries. 
apt install this one into /usr/bin, so need  `$TERMUX_PREFIX/usr/bin`

Comment (Grimler91) [reply packages/emscripten/build.sh:10]: `$TERMUX_PREFIX/usr/bin` = `/data/data/com.termux/files/usr/usr/bin`, you probably do not want /usr/usr

Comment (xeffyr) [reply packages/emscripten/build.sh:10]: Then binaryen has to be fixed.

/usr in Termux context is not available, only $TERMUX_PREFIX which stands for / and /usr at one time.

Comment (Ludea) [reply packages/emscripten/build.sh]: update to $TERMUX_PREFIX/bin

Comment (Ludea) [reply packages/emscripten/build.sh]: no need to fix binaryen, install [path ](https://github.com/termux/termux-packages/blob/master/packages/binaryen/build.sh) is right

**Comment by Ludea** [comment]
I do some test today, llvm 10.0 is not supported. I have to build and use llvm master branch.
Still need some work

File: packages/emscripten/build.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
+TERMUX_PKG_HOMEPAGE=https://emscripten.org/
+TERMUX_PKG_DESCRIPTION="Emscripten is a complete compiler toolchain to WebAssembly"
+TERMUX_PKG_LICENSE="MIT"
+TERMUX_PKG_VERSION=2.0.9
+TERMUX_PKG_SRCURL=https://github.com/emscripten-core/emscripten/archive/${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=9e54b33da5025270221b38fdc603191b8f22afa038b72c6dcfaa6423888746f5
+TERMUX_PKG_DEPENDS="python, llvm, nodejs, binaryen"
```
Comment (buttaface) [inline packages/emscripten/build.sh:7]: Is node.js needed to run this package or just to build it?

File: packages/emscripten/build.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+	./emcc -v
+}
+
+termux_step_make_install() {
+	rm $TERMUX_PREFIX/bin/emscripten -rf
+	mkdir $TERMUX_PREFIX/bin/emscripten
+	cp $TERMUX_PKG_SRCDIR/emcc $TERMUX_PREFIX/bin/emscripten/emcc
+	cp $TERMUX_PKG_SRCDIR/emcc.py $TERMUX_PREFIX/bin/emscripten/emcc.py
+	cp $TERMUX_PKG_SRCDIR/emscripten.py $TERMUX_PREFIX/bin/emscripten/emscripten.py
+	cp -r $TERMUX_PKG_SRCDIR/tools $TERMUX_PREFIX/bin/emscripten/tools
```
Comment (buttaface) [inline packages/emscripten/build.sh:20]: Could be made into one line, `cp -r $TERMUX_PKG_SRCDIR/{emcc,emcc.py,emscripten.py,tools} $TERMUX_PREFIX/bin/emscripten`. Also, these won't be in the user's path after package installation, intentional?

File: packages/emscripten/build.sh (hunk 2 above)
Comment (Ludea) [reply packages/emscripten/build.sh:7]: there is no need to build emscripten, it's a python script. And it's also need node.js

**Comment by Ludea** [comment]
Emscripten need llvm, based on master branch. I see 2 options
1. Create a subpackage into emscripten 
2. Update existing llvm package.

WDYT ?
The easy way is to update llvm

**Comment by buttaface** [comment]
It needs master or the latest 11.0 release?

**Comment by Ludea** [comment]
https://github.com/emscripten-core/emsdk/blob/master/emsdk_manifest.json#L5
It needs master

File: packages/emscripten/build.sh (hunk 3 above)
Comment (Ludea) [reply packages/emscripten/build.sh:20]: yes, I move them into bin, to be accessible to env variable path

**Comment by buttaface** [comment]
Is that just needed by the master branch of emscripten or the 2.0.9 release you're shipping here? If both, I don't think we're going to release another LLVM master snapshot package, so you could build it in this package instead if you want, [as I do with the forked LLVM that Swift uses](https://github.com/termux/termux-packages/blob/master/packages/swift/build.sh#L35).

**Comment by Ludea** [comment]
had some issue when building llvm

**Comment by Ludea** [comment]
@buttaface I get https://reviews.llvm.org/D85706?id=286724 issue when using the docker script.
I don't see you use https://github.com/termux/termux-packages/blob/master/packages/libllvm/projects-compiler-rt-CMakeLists.txt.patch into https://github.com/termux/termux-packages/tree/master/packages/Swift.

You don't get this issue ?
You fix it by another way ?

**Comment by buttaface** [comment]
The Swift package doesn't cross-compile compiler-rt, but I just saw that issue myself when natively compiling Swift trunk on my phone. What's strange is that it didn't happen a month ago when I first compiled the same source, so maybe this is a recent CMake change causing this?

You could avoid building compiler-rt if you don't use it, or apply that upstream patch you linked to your trunk build if you do.

**Comment by buttaface** [comment]
Btw, I will update the libllvm package to 11.0 this week, so I may hit this same problem and will let you know.

**Comment by buttaface** [comment]
I have not seen that build problem with LLVM 11.0, #6154, could be a more recent regression in LLVM trunk.

**Comment by truboxl** [comment]
I believe this PR should be associated with #5043

**Pull request closed without merging**